Tinkerpop 1977 - Sasl Authentication

mattallenuk opened the pull request:
Have implemented Sasl Authentication after discussion on Gremlin Users group. I think this is along the lines of the thinking of the TINKERPOP-1977 (https://issues.apache.org/jira/browse/TINKERPOP-1977). Have added an authenticator base class and implemented a SaslAuthenticator class on top of that. 

Have made modification to the DriverRemoteConnection class to accommodate the authentication class. Have also moved UUID function into utils as used in the SaslAuthenticator as well.

Tested changes on a local project worked great.

jorgebay commented on gremlin-javascript/src/main/javascript/gremlin-javascript/lib/driver/authenticator.js:
```diff
@@ -0,0 +1,32 @@
+    this._credentials = credentials;
+  }
+  
+  evaluateChallenge(ws, header) {
```
We should detach the `Authenticator` from the transport (`ws`).

I think we can provide an interface that takes a server challenge and returns a client challenge, very similar to java's counter part: https://github.com/apache/tinkerpop/blob/master/gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/auth/Authenticator.java#L55

In our case, `evaluateResponse()` should be asynchronous and return a `Promise` as some SASL mechanisms (like Kerberos) require IO on the client side.

The connection will still be responsible to build the requests and parse the response with those challenges.

mattallenuk commented:
@jorgebay. I did actually do it this way originally but was trying to avoid changing DriverRemoteConnection too much. Will upload these changes this week for review. 

@spmallette, your most welcome. Glad I can try to make some contributions.

mattallenuk commented:
@jorgebay, I've abstracted out the transport and made some changes to the DriverRemoteConnection class to handle the authenticator response. 

My only concern with the implementation is that there is a potential memory leak as there is a handler added to the DriverRemoteConnection._responseHandlers that would never get processed with sasl authentication. I had played around with adding an addHandler parameter to DriverRemoteConnection.submit() function which would by default add a handler, but not sure if this is necessary?

jorgebay commented on gremlin-javascript/src/main/javascript/gremlin-javascript/lib/driver/driver-remote-connection.js:
```diff
@@ -151,7 +156,16 @@ class DriverRemoteConnection extends RemoteConnection {
 
-    if (response.status.code >= 400) {
+    if (response.status.code === responseStatusCode.authenticationChallenge && this._authenticator) {
+      this._authenticator.evaluateChallenge(response).then(res => {
```
We can capture any possible error if we use a chained `catch()`, instead of `then()` with multiple callback functions:

```
this._authenticator.evaluateChallenge(response)
  .then(() => {
    // Any possible error thrown or promise rejection will be captured 
    return this.submit('', 'authentication', res);
  })
  .catch(handler.callback)

jorgebay commented on gremlin-javascript/src/main/javascript/gremlin-javascript/lib/driver/driver-remote-connection.js:
```diff
@@ -151,7 +156,16 @@ class DriverRemoteConnection extends RemoteConnection {
-    if (response.status.code >= 400) {
+    if (response.status.code === responseStatusCode.authenticationChallenge && this._authenticator) {
+      this._authenticator.evaluateChallenge(response).then(res => {
+        this.submit('', 'authentication', res);
```
Just a nit: we can use `null` instead of empty string to denote no `bytecode`.

jorgebay commented on gremlin-javascript/src/main/javascript/gremlin-javascript/lib/driver/remote-connection.js:
```diff
@@ -33,9 +33,11 @@ class RemoteConnection {
    * @abstract
    * @param {Bytecode} bytecode
+   * @param {String} op Operation to perform, defaults to bytecode.
+   * @param {Object} args The arguments for the operation. Defaults to
```
Missing jsdoc :)

"Defaults to an associative array containing values for "aliases" and "gremlin" keys.

mattallenuk replied:
Yep missed that, sorry :)

jorgebay commented on gremlin-javascript/src/main/javascript/gremlin-javascript/lib/driver/driver-remote-connection.js:
```diff
@@ -151,7 +156,16 @@ class DriverRemoteConnection extends RemoteConnection {
 
-    if (response.status.code >= 400) {
+    if (response.status.code === responseStatusCode.authenticationChallenge && this._authenticator) {
+      this._authenticator.evaluateChallenge(response).then(res => {
```
To avoid leaking those response handlers, you should call `_clearHandler()`:

```javascript
// Similar to how its being done below, maybe we can move that to a separate method
this._clearHandler(response.requestId);
handler.callback(null, { traversers: [] });
```

So the final solution should be something like:
```javascript
this._authenticator.evaluateChallenge(response)
  .then(() => this.submit(null, 'authentication', res))
  .then(() => this._clearHandlerAndInvokeCallback(response.requestId, handler, []))
  .catch(handler.callback);
return;
```

mattallenuk replied:
On this problem, the issue is that submit creates a new requestId for each message sent. So when sending the auth response a new requestId is created for that authentication response. The Gremlin server responds with the requestId of the previous call so the authentication requestId never gets dealt with. Am I right in thinking that your proposal is to remove all pending handlers from the handler array? Is it safe to assume that there will only ever be one handler at a time in that array, apart from when authentication is happening? Am I making sense?

jorgebay replied:
You are right, I was thinking each request have to have their own unique request id, but its not the case for auth.

The requestId that have to be sent should be the same as the one in the response.

jorgebay replied:
We should have a way to build a request (in this case an auth request) and send it providing an existing `requestId`, that doesn't add a key to the `this._responseHandlers`. That way the whole flow uses the same `requestId` and the response handler w/ the original callback is maintained until the response message (out of the auth flow) for the original request is obtained.

jorgebay commented on gremlin-javascript/src/main/javascript/gremlin-javascript/lib/driver/driver-remote-connection.js:
```diff
@@ -151,7 +156,16 @@ class DriverRemoteConnection extends RemoteConnection {
+        return handler.callback(err);
+      });
+       
+       return;
```
NIT: it looks like there is an extra spacing here :)

jorgebay commented on gremlin-javascript/src/main/javascript/gremlin-javascript/lib/driver/driver-remote-connection.js:
```diff
@@ -100,24 +105,24 @@ class DriverRemoteConnection extends RemoteConnection {
+      'op': op || 'bytecode',
       'processor': 'traversal',
-      'args': {
+      'args': args || {
```
I think each value of args map should be encoded with `this._writer.adaptObject()`

mallenuk commented:
@jorgebay. I think option b would be the better option, however, are there any plans for Tinkerpop to support anything other than "PLAIN" SASL by default. If other auth mechanisms can be configured then really full SASL support should be implemented as that will give a base for other auth mechanisms to be implemented in future. I don't know much about SASL but I can learn :)

mattallenuk commented:
@jorgebay Unfortunately I cannot get the integration tests to run locally on my mac I am getting an error on maven around gremlin-server:jar:3.4.0-snapshot not being found. I don't really know enough about maven to try and fix and cannot seem to locate anything online on how to fix :( 

Running the docker command failed on Gremlin-Server tests so didn't run the Javascript tests. 

I'll have to spend some more time trying to get the local dev env setup so that I can run the integration tests correctly.

jorgebay commented:
`submit()` accepting `requestId` looks good!

About running the test locally, you should try (on root tinkerpop directory):
- `mvn clean install -DskipTests` (run this once to generate the necessary server jars)
- `mvn verify -pl :gremlin-javascript` (run it each time you want to test)

There's more info here: http://tinkerpop.apache.org/docs/current/dev/developer/

mattallenuk commented:
Still having issues with Maven build but think I'm getting closer to making it work. Once working I'll get the integration tests working. Also going to look at option B on the Sasl class.

jorgebay commented:
We appreciate you taking the lead on this feature @mattallenuk . There is no rush, we can leave the pull request open and wait for any progress.

spmallette commented:
@mattallenuk if there are errors you can share from your build, i might be able to direct you further.

mattallenuk commented:
@spmallette, sorry, my apologies, haven't been able to look at this much over the last few days. Have some time planned nearer end of the week to progress this. I have Maven working but when I run the tests I get 2 issues.

1) On DriverRemoteConnection test I get an error: 

[ERROR] Could not deserialize the Traversal instance
java.lang.IllegalStateException: Could not locate method: GraphTraversalSource.SYNTAX_ERROR()

2) On my DriverRemoteConnectionWithSaslAuthenticator test I get a timeout:

1) DriverRemoteConnectionWithSaslAuthenticator
[INFO]        "before all" hook:
[INFO]      Uncaught Error: socket hang up
[INFO]       at TLSSocket.onHangUp (_tls_wrap.js:1124:19)
[INFO]       at endReadableNT (_stream_readable.js:974:12)
[INFO]       at _combinedTickCallback (internal/process/next_tick.js:80:11)
[INFO]       at process._tickCallback (internal/process/next_tick.js:104:9)

My code to connect to the server is as follows:

```
exports.getSecureConnectionWithAuthenticator = function getConnection(traversalSource) {
  const authenticator = new SaslAuthenticator({ username: 'stephen', password: 'password' });
  return new DriverRemoteConnection('wss://localhost:45941/gremlin', { traversalSource: traversalSource, authenticator: authenticator, rejectUnauthorized: false });
};
```

Is the problem here?

spmallette commented:
I'm not recognizing too much in the error you're seeing, but doesn't appear to be related to maven configuration or anything like that, so I probably can't help too much there.

mattallenuk commented:
I can run the tests outside of Maven with a local secure gremlin server and the tests run fine. Whenever I try to run with Maven the connection to the secure gremlin server spun-up by Maven doesn't allow my connection attempt and the web socket ends up timing out pretty much straight away. Is there anyway to debug with Maven and attempt to see why this would be happening?

mattallenuk commented:
Ok, I have found the issue. Maven doesn't create a TLS secured gremlin server for it's secure server. I changed my connection string to use a plain web socket with no TLS and the tests have run fine!

spmallette commented:
@mattallenuk glad things are working on your end. jorge is on holiday for a bit, but i'm sure he'll firm up his review as soon as he gets back.

mattallenuk commented:
@spmallette you may be able to help me with this. I'm trying to understand the SASL authentication process looking at how the Java code handles it (https://github.com/apache/tinkerpop/blob/tp33/gremlin-driver/src/main/java/org/apache/tinkerpop/gremlin/driver/Handler.java#L88-L110). In my code, I haven't pushed it yet, I am receiving from the server the 407 authorisation required request and I am sending an initial message containing { saslMechanism: "PLAIN" } but the response I get back is authentication failed. Is there something that I'm missing? Am I right in that I should be sending that message then wait for the initial response and send the sasl authentication message? I know that currently only PLAIN mechanism is accepted and I can send that without declaring what mechanism I'm sending, but I'm trying to make the code future-proof for when other authentication drivers and sasl mechanisms are supported. Any insight would be greatly appreciated.

spmallette commented:
you need to send a `sasl` key with the `saslMechanism` otherwise the message doesn't validate and you end up with an UNAUTHORIZED. Note the sending of the `NULL_CHALLENGE`

mattallenuk commented:
@spmallette thank you I'll give that a go.

mattallenuk commented:
@jorgebay I've modified the Sasl Authenticator so that a mechanism can now be passed in along with any mechanism options. I've changed the Sasl Plain Authenticator to mimic the Java handler code as discussed, however I had to do some fudging as I couldn't get the flow to work right. 

From my understanding of the Java code the Driver/Handler/GremlinSaslAuthenticationHandler class receives a 407 from the Gremlin Server and then sends a message containing { saslMechanism: 'PLAIN', sasl: (A base64 encoded null byte string) }, the server then returns an initial response and the client then sends the actual sasl authentication message... I tried that and I was getting a message that "Authentication ID must not be null". Having looked at the Gremlin Server code and the Handler/SaslAuthenticationHandler class and the Auth/SimpleAuthentication class it doesn't appear that saslMechanism token is actually parsed and the the server expects the actual sasl message right away. So, currently, in the JS SaslAuthenticator class I've passed the full Sasl message with the initial SaslMechanism argument. 

I hope that makes sense and if it needs changing I'm happy to do so.

jorgebay commented on gremlin-javascript/src/main/javascript/gremlin-javascript/lib/driver/auth/sasl-authenticator.js:
```diff
@@ -0,0 +1,48 @@
+   * @return {Object} A Promise that resolves to a valid sasl response object.
+   */
+  evaluateChallenge(challenge) {
+    return Promise.resolve(this._options.mechanism.evaluateChallenge(challenge));
```
I think we should not use `Promise.resolve()` here, we should rely on the return value of the mechanism (a `Promise`).

jorgebay commented on gremlin-javascript/src/main/javascript/gremlin-javascript/test/integration/sasl-authentication-tests.js:
```diff
@@ -0,0 +1,68 @@
+        .then(function (response) {
+          assert.ok(response);
+          assert.ok(response.traversers);
+          //assert.strictEqual(response.traversers.length, 1);
```
NIT: remove commented out lines.

mattallenuk commented:
@jorgebay Yes that is a great idea have implemented and made the changes requested. Have also updated the tests to use the PlainTextSaslAuthenticator class.

jorgebay commented:
Nice! Looks great!

VOTE: +1

spmallette commented:
I'm generally VOTE +1 but have some questions/points to make:

Is this ok pointed at master branch? any reason not to go to tp32/tp33, or is this a feature we only want on 3.4.0 on master?

also, i would say that there are three pieces of documentation i generally look for on a change likes this:

1. a [CHANGELOG](https://github.com/apache/tinkerpop/blob/ea18963e543de90668f401938217f5aa21f2ff29/CHANGELOG.asciidoc) entry
2. [upgrade documentation](https://github.com/apache/tinkerpop/blob/ea18963e543de90668f401938217f5aa21f2ff29/docs/src/upgrade/release-3.2.x-incubating.asciidoc) to call attention to this neat new feature
2. [reference documentation](https://github.com/apache/tinkerpop/blob/ea18963e543de90668f401938217f5aa21f2ff29/docs/src/reference/gremlin-variants.asciidoc#gremlin-javascript) which will explain usage in detail

At minimum, I think a change like this requires 1 and 2.  Maybe usage is self-explanatory enough to not require 3???

@jorgebay since you will be the one who will likely take this to merge, I'll let you sort out how to settle up these mostly administrative odds/ends. 

@mattallenuk thanks for being patient on this one and working through all the issues . it's appreciated. i hope that we can see more pull requests from you in the future.

mattallenuk commented:
@spmallette, @jorgebay it has been a pleasure to contribute back, thank you both for helping me get it done and for helping to improve my skills as the pull has progressed. This is my first proper pull request so it has been educating :) 

I am happy to do any of the tidying-up tasks if needed. I think there should be reference documentation just to make it very clear to people using the api.

jorgebay commented:
Great contribution @mattallenuk! 

We should merge it to `tp32`. @mattallenuk do you want to change the base branch of this pull request and rebase? There shouldn't be any conflicts. We can either rebase and [change the base branch](https://help.github.com/articles/changing-the-base-branch-of-a-pull-request/) here or I could do it before merging it. lmk what you prefer.

About the documentation entries suggested by @spmallette, do you want to include it in this pull request @mattallenuk?

mattallenuk commented:
@jorgebay I've rebased my TINKERPOP-1977 onto tp32 and made changes to the documentation. Not sure I've done it right as I'm seeing lots of conflicts here now lol! 

Let me know if I've done it wrong and how to do it correctly.

spmallette commented:
Note that it's pulling in commits from master now. Not sure how you got to that point exactly, but I think you need to just rebase your commits in your local repository on your branch.

```text
git checkout TINKERPOP-1977
git rebase origin/tp32
git push origin TINKERPOP-1977 --force
```

note that i'm assuming that your `origin/tp32` is up to date with the upstream TinkerPop `tp32` and thus the `rebase` command should take your commits and re-write them on top of the latest commits in the `tp32` branch.  Before you `push`, please review `git log` and make sure that you only see your commits followed by the TinkerPop history from `tp32`.  If something still looks funny there, just ping back and we can figure it out.

Once you do all this, the PR should update automatically and you should only see your commits on it and a "Files changed" of the same.

mattallenuk commented:
@spmallette thanks I think I've got it right now. Not a git expert, still learning :)

jorgebay commented:
I think the git branch is still based on`master` instead of `tp32`, that's why github shows a lot of additional commits.

spmallette commented:
hmm - something still looks fishy:

https://github.com/mattallenuk/tinkerpop/commits/TINKERPOP-1977

we should have a string of your continuous commits on top of `tp32` but they are instead interspersed with the other stuff. not sure how it got in that state. 

let me think about how to `git` your way out of this and get the commit history clean.

mattallenuk commented:
:( sorry

mattallenuk commented:
If i look at https://github.com/mattallenuk/tinkerpop/commits/tp32, it appears to be what you were expecting? Have I done it the wrong way around?

jorgebay commented:
its not a problem :) 

As long as you make sure you don't loose your commits on the process (make sure you push a backup branch to your personal repository).

I think the issue is that your `tp32` branch is incorrect (looks like master): https://github.com/apache/tinkerpop/compare/tp32...mattallenuk:tp32

First you have to restore your `tp32` branch locally.

What's your output for: `git remote -vv` ?

I think that with that information, we will be able to help you all the way.

mattallenuk commented:
origin  https://github.com/mattallenuk/tinkerpop.git (fetch)
origin  https://github.com/mattallenuk/tinkerpop.git (push)
upstream        https://github.com/apache/tinkerpop.git (fetch)
upstream        https://github.com/apache/tinkerpop.git (push)

jorgebay commented:
awesome!

First, lets restore `tp32` on your repo:

```
git fetch upstream tp32
git checkout tp32
git reset --hard upstream/tp32
git push origin tp32
```

Once you did that, you can try to rebase.

mattallenuk commented:
Ok, done that which way do I rebase? Tried 

```
git checkout TINKERPOP-1977
git rebase origin/tp32
```

and getting some merges required with files I've not touched.

mattallenuk commented:
```
First, rewinding head to replay your work on top of it...
Generating patches: 100% (757/757), done.
Applying: valueMap should always return string keys
Using index info to reconstruct a base tree...
M       gremlin-groovy-test/src/main/groovy/org/apache/tinkerpop/gremlin/process/traversal/step/map/GroovyValueMapTest.groovy
M       gremlin-test/src/main/java/org/apache/tinkerpop/gremlin/process/computer/GraphComputerTest.java
M       gremlin-test/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/ValueMapTest.java
.git/rebase-apply/patch:33: space before tab in indent.
                // add tokens, as string keys
.git/rebase-apply/patch:57: trailing whitespace.

.git/rebase-apply/patch:60: space before tab in indent.
                new ScriptTraversal<>(g, "gremlin-groovy", "g.V().hasLabel('person').filter(__.outE('created')).valueMap(true)")
.git/rebase-apply/patch:72: trailing whitespace.

.git/rebase-apply/patch:135: trailing whitespace.
     *
warning: squelched 19 whitespace errors
warning: 24 lines add whitespace errors.
Falling back to patching base and 3-way merge...
Auto-merging gremlin-test/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/ValueMapTest.java
CONFLICT (content): Merge conflict in gremlin-test/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/ValueMapTest.java
Auto-merging gremlin-test/src/main/java/org/apache/tinkerpop/gremlin/process/computer/GraphComputerTest.java
CONFLICT (content): Merge conflict in gremlin-test/src/main/java/org/apache/tinkerpop/gremlin/process/computer/GraphComputerTest.java
Auto-merging gremlin-groovy-test/src/main/groovy/org/apache/tinkerpop/gremlin/process/traversal/step/map/GroovyValueMapTest.groovy
error: Failed to merge in the changes.
Patch failed at 0001 valueMap should always return string keys
Use 'git am --show-current-patch' to see the failed patch

Resolve all conflicts manually, mark them as resolved with
"git add/rm <conflicted_files>", then run "git rebase --continue".
You can instead skip this commit: run "git rebase --skip".
To abort and get back to the state before "git rebase", run "git rebase --abort".
```

jorgebay commented:
At this stage you have to use "rebase interactive" to manually select your commits and skip the rest.

If you have a tool to do that, it would make your job a lot easier.

mattallenuk commented:
Ah, thinking I've made a complete hash of this. My git rebase knowledge is woeful lol.

I ran `git reset --hard ORIG_HEAD` on my local repo to get back to pre-rebase. 

I've tried rebasing onto origin/tp32 via cli and SourceTree. When I try to interactive rebase on the cli I get a list of 735 commits it wants to apply. The rebase commands shown don't have a skip so am I supposed to just delete all these commits and keep the ones I'm interested in? There is a warning message that these commits will be lost if I do that...

As I said my rebase knowledge is not great and having read a bunch of articles on rebasing I am none the wiser. Sorry to be a pain guys, hopefully with your guidance I can get this sorted.

mattallenuk commented:
@jorgebay I think I've got it now. Starting to see through the fog. Hopefully that last push is the one :)

jorgebay commented:
Its looking great!

I'll merge it to a dev branch into the project repo and fix the minor test failure.

jorgebay commented:
Merged 🎉! Thanks @mattallenuk for your contribution.

mattallenuk commented:
Awesome, thanks guys for the help in getting it there!
